### PR TITLE
KAFKA-7737; Use single path in producer for initializing the producerId

### PR DESCRIPTION
**Pull request by hachikuji**
Previously the idempotent producer and transactional producer use separate logic when initializing the producerId. This patch consolidates the two paths. We also do some cleanup in `TransactionManagerTest` to eliminate brittle expectations on `Sender`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by hachikuji** [comment]
retest this please

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java [hunk 1]
```diff
@@ -517,16 +512,6 @@ public boolean isRunning() {
         return running;
     }
 
-    private ClientResponse sendAndAwaitInitProducerIdRequest(Node node) throws IOException {
```
Comment (viktorsomogyi) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:520]: Just sanity checking: do I understand correctly that this became the part of `resetProducerIdIfNeeded` so at the first time this code path runs the producer id gets placed in the transactional request queue? Is this also the reason for removing 'maybeWaitForProducerId`?

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 2]
```diff
@@ -2925,4 +2766,14 @@ private void assertFutureFailed(Future<RecordMetadata> future) throws Interrupte
         }
     }
 
+    private void runUntil(Supplier<Boolean> condition) {
```
Comment (viktorsomogyi) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:2984]: Would you consider using a `long timeout` instead and move this to TestUtils? Perhaps we need this somewhere else too.

**Review by viktorsomogyi — APPROVED** [review]
It looks good as far as I can tell. Saw that you had 2 test issues but most likely those were flakies, I reran the client and core tests for this patch on my laptop and they didn't fail, although let's rerun them in the hope of a green build.

**Comment by viktorsomogyi** [comment]
Ah you reran them in the meanwhile :)
Now only one failed: 
`kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup` for which we have a submitted issue (KAFKA-7965)

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java (hunk 1 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:520]: Yes, that's right.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java (hunk 2 above)
Comment (hachikuji) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:2984]: It's a bit difficult to restructure these test cases with a timeout. Most of the logic is basically "run the sender until X happens, then verify Y." With a timeout, we cannot control the stopping condition.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java [hunk 3]
```diff
@@ -538,41 +523,6 @@ private Node awaitNodeReady(FindCoordinatorRequest.CoordinatorType coordinatorTy
         return null;
     }
 
-    private void maybeWaitForProducerId() {
-        while (!forceClose && !transactionManager.hasProducerId() && !transactionManager.hasError()) {
-            Node node = null;
-            try {
-                node = awaitNodeReady(null);
-                if (node != null) {
-                    ClientResponse response = sendAndAwaitInitProducerIdRequest(node);
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:547]: The only other caller of `InitProducerIdRequest.Builder` is in TransactionManager#initializeTransactions which is only called in producer#iniTxns. For idempotent producer how that initPid request would be sent? Did I miss anything?

**Review by guozhangwang — COMMENTED** [review]
Just had one qq about the PR, will review the remaining of the PR once that's cleared -- maybe I missed something big here so need to clarify.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java (hunk 3 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:547]: We also build an `InitProducerIdRequest` inside `TransactionManager.resetProducerIdIfNeeded` in this patch.

Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:547]: Ack, thanks!

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java [hunk 4]
```diff
@@ -481,6 +489,17 @@ synchronized void resetProducerIdIfNeeded() {
         if (shouldResetProducerStateAfterResolvingSequences())
             // Check if the previous run expired batches which requires a reset of the producer state.
             resetProducerId();
+
+        if (!isTransactional()
+                && currentState != State.INITIALIZING
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java]: qq: is the second condition necessary given the first one? If we are not in transactional, then right now the only place to transit to INITIALIZING is line 496 below. I'm actually fine to leave it here to be less bug-vulnerable, but just to clarify my understanding.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java [hunk 5]
```diff
@@ -300,28 +295,27 @@ public void run() {
...
-                            "some previously sent messages and can no longer retry them. It isn't safe to continue."));
-                } else if (maybeSendAndPollTransactionalRequest()) {
-                    return;
+                            new KafkaException("The client hasn't received acknowledgment for " +
+                                    "some previously sent messages and can no longer retry them. It isn't safe to continue."));
                 }
 
-                // do not continue sending if the transaction manager is in a failed state or if there
-                // is no producer id (for the idempotent case).
-                if (transactionManager.hasFatalError() || !transactionManager.hasProducerId()) {
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:318]: qq: could you add a comment explaining why we need to call `client.poll` before exiting under this condition?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java [hunk 6]
```diff
@@ -481,6 +489,17 @@ synchronized void resetProducerIdIfNeeded() {
         if (shouldResetProducerStateAfterResolvingSequences())
             // Check if the previous run expired batches which requires a reset of the producer state.
             resetProducerId();
+
+        if (!isTransactional()
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java]: We can let this to return a boolean on whether a request is enqueued, and then in Sender we could:

```
if (transactionManager.resetProducerIdIfNeeded() && maybeSendAndPollTransactionalRequest()) ..
```

also nit: Add a comment on top indicating that we will send out a new initPid request if producer is not transactional.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java [hunk 7]
```diff
@@ -412,6 +416,10 @@ synchronized void transitionToAbortableError(RuntimeException exception) {
 
     synchronized void transitionToFatalError(RuntimeException exception) {
         transitionTo(State.FATAL_ERROR, exception);
+
+        if (pendingResult != null) {
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:420]: qq: is that a piggy-backed fix, or is it necessary for the refactoring?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java [hunk 8]
```diff
@@ -216,7 +216,7 @@ public void resetSequenceNumbers(Consumer<ProducerBatch> resetSequence) {
         private boolean isTransitionValid(State source, State target) {
             switch (target) {
                 case INITIALIZING:
-                    return source == UNINITIALIZED;
+                    return source == UNINITIALIZED || source == READY;
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java]: How about: letting the state to transit to `UNINITIALIZED` inside `resetProducerId` (since it will only execute successfully if not transactional). By doing this we still have a single transition path to INITIALIZING?

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java [hunk 9]
```diff
@@ -1154,7 +1154,9 @@ public void testExpiryOfAllSentBatchesShouldCauseUnresolvedSequences() throws Ex
         Deque<ProducerBatch> batches = accumulator.batches().get(tp0);
         assertEquals(0, batches.size());
         assertTrue(transactionManager.hasProducerId(producerId));
+
         // We should now clear the old producerId and get a new one in a single run loop.
+        time.sleep(10);
```
Comment (guozhangwang) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:1160]: qq: Why is this necessary?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java (hunk 4 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java]: I added it to ensure that we don't enqueue multiple InitProducerId requests.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java (hunk 6 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java]: I think we want to call `maybeSendAndPollTransactionalRequest` in this loop regardless whether `resetProducerIdNeeded` returns true. Ack on the comment.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java (hunk 7 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:420]: More of an optimization than a fix I guess, but it simplified one of the tests in `TransactionManagerTest`.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java (hunk 8 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java]: Yeah, it's a good thought. I don't recall if I previously considered this, but let me take a look and see if it works.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java (hunk 9 above)
Comment (hachikuji) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:1160]: Note that we blackout the node for 10ms above. Previously we were relying on the backoff logic in `maybeWaitForProducerId` for the node to be ready again. Now the test needs time to be advanced externally since we cannot rely on `client.poll` advancing it.

Comment (guozhangwang) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:1160]: SG.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java (hunk 5 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:318]: I don't have a great explanation for it to be honest. I guess we are trying to ensure that any inflight requests return before we shutdown. Perhaps we should just crash the thread and let the client get cleaned up? How about I open a separate issue for this?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java [hunk 10]
```diff
@@ -451,7 +453,7 @@ boolean hasProducerIdAndEpoch(long producerId, short producerEpoch) {
     /**
      * Set the producer id and epoch atomically.
      */
-    void setProducerIdAndEpoch(ProducerIdAndEpoch producerIdAndEpoch) {
+    private void setProducerIdAndEpoch(ProducerIdAndEpoch producerIdAndEpoch) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:457]: Reducing visibility here to make sure that test cases are forced through the proper state transitions.

**Comment by hachikuji** [comment]
retest this please

**Comment by hachikuji** [comment]
retest this please

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java (hunk 5 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:318]: SG.

**Review by guozhangwang — APPROVED** [review]
LGTM!

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java [hunk 11]
```diff
@@ -473,19 +475,32 @@ void setProducerIdAndEpoch(ProducerIdAndEpoch producerIdAndEpoch) {
...
+        transitionTo(State.UNINITIALIZED);
     }
 
-    synchronized void resetProducerIdIfNeeded() {
-        if (shouldResetProducerStateAfterResolvingSequences())
-            // Check if the previous run expired batches which requires a reset of the producer state.
-            resetProducerId();
+    synchronized void resetIdempotentProducerIdIfNeeded() {
+        if (!isTransactional()) {
+            if (shouldResetProducerStateAfterResolvingSequences()) {
```
Comment (bob-barrett) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:491]: Just FYI, for KIP-360 I'm doing this check for both idempotent and transactional, since it triggers an epoch bump instead of a producerId reset. I'll just pull this call out to a shared code path, the rest of this method shouldn't need to change.

**Review by bob-barrett — APPROVED** [review]
LGTM

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java (hunk 11 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:491]: Sounds good. We can rename the method as well.

**Comment by hachikuji** [comment]
retest this please
